From 095c97734720d41e034ad0d2b07c2faf0ce22517 Mon Sep 17 00:00:00 2001
From: Michal Srb <michalsrb@gmail.com>
Date: Tue, 21 Feb 2017 23:15:10 +0200
Subject: [PATCH 08/13] Add FcPreprocessFilename speeding up FcCompareFilename.

FcPreprocessFilename calculates hash of the filename to fast reject not matching filenames. Filenames are compared both case sensitive and case insensitive. The hash is calculated for case insensitive variant. That way if hash is not matching, we know for sure that both kinds of string comparison would fail.

In adition FcPreprocessFilename remembers whether the filename contains any globs ('?', '*') so that FcCompareFilename cam quickly skip over FcStrGlobMatch in case there are no globs.

TODO: Add benchmark numbers.
---
 src/fcmatch.c | 33 ++++++++++++++++++++++++++++++---
 src/fcobjs.h  |  2 +-
 2 files changed, 31 insertions(+), 4 deletions(-)

diff --git a/src/fcmatch.c b/src/fcmatch.c
index 4e95a86..b81616f 100644
--- a/src/fcmatch.c
+++ b/src/fcmatch.c
@@ -265,15 +265,42 @@ bail:
     return ret;
 }
 
+static FcPrepValue
+FcPreprocessFilename (FcValue *v)
+{
+    FcPrepValue prep;
+
+    prep.type = FcPrepStrFilename;
+
+    const FcChar8* str = FcValueString(v);
+    prep.str_hash = FcStrHashIgnoreBlanksAndCase(str);
+    prep.filename_has_globs = (strchr((const char *)str, '*') != NULL && strchr((const char *)str, '?') != NULL);
+
+    return prep;
+}
+
 static double
 FcCompareFilename (FcValue *v1, FcPrepValue *p1, FcValue *v2, FcPrepValue *p2)
 {
+    FcBool may_have_globs = FcTrue;
+    FcBool may_be_same = FcTrue;
+
+    if (p1->type == FcPrepStrFilename)
+    {
+	may_have_globs = p1->filename_has_globs;
+
+	if (p2->type == FcPrepStrFilename)
+	{
+	    may_be_same = (p1->str_hash == p2->str_hash);
+	}
+    }
+
     const FcChar8 *s1 = FcValueString (v1), *s2 = FcValueString (v2);
-    if (FcStrCmp (s1, s2) == 0)
+    if (may_be_same && FcStrCmp (s1, s2) == 0)
 	return 0.0;
-    else if (FcStrCmpIgnoreCase (s1, s2) == 0)
+    else if (may_be_same && FcStrCmpIgnoreCase (s1, s2) == 0)
 	return 1.0;
-    else if (FcStrGlobMatch (s1, s2))
+    else if (may_have_globs && FcStrGlobMatch (s1, s2))
 	return 2.0;
     else
 	return 3.0;
diff --git a/src/fcobjs.h b/src/fcobjs.h
index 3b1a2e8..4cb8cd8 100644
--- a/src/fcobjs.h
+++ b/src/fcobjs.h
@@ -42,7 +42,7 @@ FC_OBJECT (HINTING,		FcTypeBool,	NULL,			NULL)
 FC_OBJECT (VERTICAL_LAYOUT,	FcTypeBool,	NULL,			NULL)
 FC_OBJECT (AUTOHINT,		FcTypeBool,	NULL,			NULL)
 FC_OBJECT (GLOBAL_ADVANCE,	FcTypeBool,	NULL,			NULL)	/* deprecated */
-FC_OBJECT (FILE,		FcTypeString,	FcCompareFilename,	NULL)
+FC_OBJECT (FILE,		FcTypeString,	FcCompareFilename,	FcPreprocessFilename)
 FC_OBJECT (INDEX,		FcTypeInteger,	NULL,			NULL)
 FC_OBJECT (RASTERIZER,		FcTypeString,	FcCompareString,	FcPreprocessString)	/* deprecated */
 FC_OBJECT (OUTLINE,		FcTypeBool,	FcCompareBool,		NULL)
-- 
2.15.1

